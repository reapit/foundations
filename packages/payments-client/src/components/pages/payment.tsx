import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router'
import useSWR from 'swr'
import {
  Section,
  H3,
  Helper,
  Loader,
  Formik,
  Form,
  CardInputGroup,
  Input,
  LevelRight,
  Button,
  unformatCard,
  unformatCardExpires,
  H5,
  H6,
  Grid,
  GridItem,
} from '@reapit/elements'
import { PaymentModel } from '@reapit/foundations-ts-definitions'
import { URLS } from '../../constants/api'
import { MerchantKey, opayoMerchantKeyService } from '../../opayo-api/merchant-key'
import { opayoCreateTransactionService } from '../../opayo-api/transactions'
import PaymentCustomerSection from '../ui/payments/payment-customer-section'
import PropertySection from '../ui/payments/property-section'

export const PaymentForm: React.FC<{ data: PaymentModel; merchantKey: MerchantKey }> = ({ data, merchantKey }) => {
  const customer = data.customer || {}
  const { forename = '', surname = '', primaryAddress = {} } = customer
  const { buildingName, buildingNumber, line1, line3, line4, postcode = '', countryId = '' } = primaryAddress
  let address1: string
  if (buildingName && line1) {
    address1 = `${buildingName} ${line1}`
  } else if (buildingNumber && line1) {
    address1 = `${buildingNumber} ${line1}`
  } else {
    address1 = buildingName || buildingNumber || line1 || ''
  }

  return (
    <Formik
      initialValues={{
        customerFirstName: forename,
        customerLastName: surname,
        address1,
        city: line3 || line4 || '',
        postalCode: postcode,
        country: countryId,
        cardholderName: '',
        cardNumber: '',
        expiryDate: '',
        securityCode: '',
        cardIdentifier: '',
      }}
      onSubmit={cardDetails => {
        const {
          cardholderName,
          cardNumber,
          expiryDate,
          securityCode,
          customerFirstName,
          customerLastName,
          address1,
          city,
          postalCode,
          country,
        } = cardDetails
        window
          .sagepayOwnForm({
            merchantSessionKey: merchantKey.merchantSessionKey,
          })
          .tokeniseCardDetails({
            cardDetails: {
              cardholderName,
              cardNumber: unformatCard(cardNumber),
              expiryDate: unformatCardExpires(expiryDate),
              securityCode,
            },
            onTokenised: async (result: any) => {
              if (result.success) {
                const payment = await opayoCreateTransactionService({
                  transactionType: 'Payment',
                  paymentMethod: {
                    card: {
                      merchantSessionKey: merchantKey.merchantSessionKey,
                      cardIdentifier: result.cardIdentifier,
                      save: false,
                    },
                  },
                  vendorTxCode: `demotransaction-${Math.floor(Math.random() * 1000)}`,
                  amount: 1000,
                  currency: 'GBP',
                  description: 'Rent payment',
                  apply3DSecure: 'Disable',
                  customerFirstName,
                  customerLastName,
                  billingAddress: {
                    address1,
                    city,
                    postalCode,
                    country,
                  },
                  entryMethod: 'Ecommerce',
                })

                console.log('payment result is', payment)
              } else {
                console.log(JSON.stringify(result))
              }
            },
          })
      }}
    >
      {() => (
        <Form className="form">
          <CardInputGroup hasBillingAddress whiteListTestCards={['4929000000006']} />
          <Section>
            <Input id="cardIdentifier" type="hidden" name="cardIdentifier" />
            <LevelRight>
              <Button variant="primary" type="submit">
                Make Payment
              </Button>
            </LevelRight>
          </Section>
        </Form>
      )}
    </Formik>
  )
}

const PaymentPage: React.FC = () => {
  const { paymentId } = useParams<{ paymentId: string }>()
  const { data } = useSWR<PaymentModel>(`${URLS.PAYMENTS}/${paymentId}`)
  const [loading, setLoading] = useState(false)
  const [merchantKey, setMerchantKey] = useState<MerchantKey>()

  useEffect(() => {
    const fetchmerchantKey = async () => {
      const fetchedKey = await opayoMerchantKeyService()
      if (fetchedKey) {
        setMerchantKey(fetchedKey)
      }
      setLoading(false)
    }

    fetchmerchantKey()

    setLoading(true)
  }, [setMerchantKey])

  if (loading) {
    return <Loader />
  }

  const { customer, amount, description, externalReference, propertyId } = data || {}
  return (
    <>
      <H3 isHeadingSection>Card Payment</H3>
      <H5 isHeadingSection>Payment For {paymentId}</H5>
      {!merchantKey && (
        <Helper variant="info">
          Welome to Reapit Payments portal. It seems you dont currently have an account registered with Opayo. Please
          talk to your administrator to set this up for you.
        </Helper>
      )}
      <Section>
        <Grid>
          <PaymentCustomerSection customer={customer} />
          <PropertySection propertyId={propertyId} />
          <GridItem>
            <H5>GBP: {amount}</H5>
            <H6>Payment description: {description}</H6>
            <H6>Payment ref: {externalReference}</H6>
          </GridItem>
        </Grid>
        <H6>Payment amount: {amount} GBP</H6>
      </Section>
      {merchantKey && data ? <PaymentForm data={data} merchantKey={merchantKey} /> : <Loader />}
    </>
  )
}

export default PaymentPage
