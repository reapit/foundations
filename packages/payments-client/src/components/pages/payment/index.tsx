import React, { FC } from 'react'
import { useLocation } from 'react-router'
import PrivateRouteWrapper from '../../../core/private-route-wrapper'
import { PaymentsProvider } from '../../../core/use-payments-state'
import { MainContainer } from '@reapit/elements'
import ErrorBoundary from '../../../core/error-boundary'
import Payment from './payment'

export const PaymentPage: FC = () => {
  const location = useLocation()
  const { search } = location
  const queryParams = new URLSearchParams(search)
  const session = queryParams.get('session')
  const clientId = queryParams.get('clientCode')

  if (session && clientId) {
    return (
      <PaymentsProvider>
        <ErrorBoundary>
          <MainContainer>
            <Payment />
          </MainContainer>
        </ErrorBoundary>
      </PaymentsProvider>
    )
  }

  return (
    <ErrorBoundary>
      <PrivateRouteWrapper>
        <Payment />
      </PrivateRouteWrapper>
    </ErrorBoundary>
  )
}

export default PaymentPage
