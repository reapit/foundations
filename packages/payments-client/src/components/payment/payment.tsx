import React, { useState, useEffect, FC, Dispatch, SetStateAction } from 'react'
import { useReapitConnect } from '@reapit/connect-session'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { PaymentModel, PropertyModel } from '@reapit/foundations-ts-definitions'
import { elFadeIn, Loader, PersistentNotification } from '@reapit/elements'
import {
  ClientConfigModel,
  CreateTransactionModel,
  MerchantKey,
  PaymentEmailReceipt,
  PaymentPageContent,
  PaymentProvider,
  PaymentProviderInitialisers,
  ThreeDSecureResponse,
  UpdateStatusBody,
} from '@reapit/payments-ui'
import {
  GetActionNames,
  getActions,
  useReapitGet,
  useReapitUpdate,
  updateActions,
  UpdateActionNames,
  UpdateReturnTypeEnum,
  SendFunction,
  ReapitGetState,
} from '@reapit/use-reapit-data'
import { useConfigState } from '../../core/use-config-state'
import dayjs from 'dayjs'

export interface PaymentProps {
  paymentRequest: ReapitGetState<PaymentModel>
}

export interface SetProviderParams extends Omit<PaymentProviderInitialisers, 'config' | 'payment' | 'merchantKey'> {
  paymentProvider: PaymentProvider | null
  setPaymentProvider: Dispatch<SetStateAction<PaymentProvider | null>>
  config: ClientConfigModel | null
  payment: PaymentModel | null
  merchantKey?: MerchantKey
}

export const handleSetProvider =
  ({ config, payment, property, merchantKey, paymentProvider, setPaymentProvider, ...rest }: SetProviderParams) =>
  () => {
    const paymentHasProperty = payment?.propertyId
    const propertyFetched = !paymentHasProperty || (paymentHasProperty && property)
    const paymentHasChanged =
      paymentProvider?.payment?._eTag && payment?._eTag && paymentProvider.payment._eTag !== payment._eTag
    const merchantKeyHasChanged =
      paymentProvider?.merchantKey && merchantKey?.expiry && paymentProvider?.merchantKey.expiry !== merchantKey?.expiry
    const providerInvalid =
      !paymentProvider || (paymentProvider && paymentHasChanged) || (merchantKey && merchantKeyHasChanged)

    if (config && payment && propertyFetched && merchantKey && providerInvalid) {
      const paymentProvider = new PaymentProvider({
        config,
        payment,
        property,
        merchantKey,
        ...rest,
      })
      setPaymentProvider(paymentProvider)
    }
  }

export const handleGetMerchantKey =
  (getMerchantKey: SendFunction<void, boolean | MerchantKey>, idToken: string, config: ClientConfigModel | null) =>
  () => {
    if (idToken && config?.clientCode && config?.isConfigured) {
      getMerchantKey()
    }
  }

export const handleMerchantKeyRefresh =
  (
    merchantKey: MerchantKey | undefined,
    getMerchantKey: SendFunction<void, boolean | MerchantKey>,
    payment: PaymentModel | null,
    config: ClientConfigModel | null,
  ) =>
  () => {
    if (!merchantKey || !config?.isConfigured) return

    const expiry = dayjs(merchantKey.expiry)

    const timer = setInterval(() => {
      if (expiry.isBefore(dayjs().add(1, 'minute'))) {
        clearInterval(timer)
        getMerchantKey()
      }
    }, 45000)

    if (payment?.status === 'posted') {
      clearInterval(timer)
    }

    return () => {
      clearInterval(timer)
    }
  }

export const handleDeleteNomTran = () => () => {
  const nomTranCode = window['__REAPIT_MARKETPLACE_GLOBALS__'] && window['__REAPIT_MARKETPLACE_GLOBALS__'].nomTranCode
  if (nomTranCode) {
    window['__REAPIT_MARKETPLACE_GLOBALS__'].nomTranCode = null
  }
}

export const Payment: FC<PaymentProps> = ({ paymentRequest }) => {
  const [paymentProvider, setPaymentProvider] = useState<PaymentProvider | null>(null)
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { config, configLoading } = useConfigState()
  const configNotConfigured = !config?.isConfigured

  const clientCode = connectSession?.loginIdentity?.clientId
  const idToken = connectSession?.idToken ?? ''

  const [payment, paymentLoading, , refreshPayment] = paymentRequest
  const paymentId = payment?.id

  const propertyId = payment?.propertyId

  const [property, propertyLoading] = useReapitGet<PropertyModel>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getPropertyById],
    uriParams: {
      propertyId,
    },
    fetchWhenTrue: [propertyId],
  })

  const [receiptLoading, , receiptSubmit] = useReapitUpdate<PaymentEmailReceipt, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.submitPrivatePaymentReceipt],
    method: 'POST',
    headers: {
      Authorization: idToken,
      'reapit-customer': clientCode as string,
      'reapit-app-id': process.env.appId,
    },
    uriParams: {
      paymentId,
    },
  })

  const [statusLoading, , statusSubmit] = useReapitUpdate<UpdateStatusBody, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.privatePaymentUpdate],
    method: 'PATCH',
    headers: {
      'if-match': payment?._eTag as string,
    },
    uriParams: {
      paymentId,
    },
  })

  const [merchantKeyLoading, merchantKey, getMerchantKey] = useReapitUpdate<void, MerchantKey>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.getOpayoMerchantKey],
    method: 'POST',
    returnType: UpdateReturnTypeEnum.RESPONSE,
    headers: {
      Authorization: idToken,
      'reapit-customer': clientCode as string,
      'reapit-app-id': process.env.appId,
    },
  })

  const [, , transactionSubmit] = useReapitUpdate<CreateTransactionModel, ThreeDSecureResponse>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.submitOpayoTransaction],
    method: 'POST',
    returnType: UpdateReturnTypeEnum.RESPONSE,
    headers: {
      Authorization: idToken,
      'reapit-customer': clientCode as string,
      'reapit-app-id': process.env.appId,
    },
  })

  useEffect(handleGetMerchantKey(getMerchantKey, idToken, config), [config, idToken])

  useEffect(handleMerchantKeyRefresh(merchantKey, getMerchantKey, payment, config), [merchantKey, payment, config])

  useEffect(handleDeleteNomTran(), [])

  useEffect(
    handleSetProvider({
      paymentProvider,
      setPaymentProvider,
      config,
      payment,
      property,
      merchantKey,
      getMerchantKey,
      receiptAction: {
        receiptLoading,
        receiptSubmit,
      },
      statusAction: {
        statusLoading,
        statusSubmit,
      },
      transactionSubmit: transactionSubmit as SendFunction<CreateTransactionModel, ThreeDSecureResponse>,
      refreshPayment,
      isPortal: false,
    }),
    [
      paymentProvider,
      config,
      payment,
      property,
      merchantKey,
      receiptSubmit,
      statusSubmit,
      transactionSubmit,
      refreshPayment,
    ],
  )

  if (configNotConfigured && !configLoading) {
    return (
      <PersistentNotification className={elFadeIn} intent="danger" isFullWidth isInline isExpanded>
        The app cannnot currently process client payments. This is likely because your payment provider has not been
        configured. Please contact your Reapit Organisation Administrator or if you are an Admin, use the dedicated page
        in the main navigation.
      </PersistentNotification>
    )
  }

  if ((configLoading || paymentLoading || propertyLoading || merchantKeyLoading) && !paymentProvider) {
    return <Loader />
  }

  if (!payment) {
    return (
      <PersistentNotification className={elFadeIn} intent="primary" isFullWidth isInline isExpanded>
        We do not have any information about this payment.
      </PersistentNotification>
    )
  }

  if (!paymentProvider) {
    return (
      <PersistentNotification className={elFadeIn} intent="danger" isFullWidth isInline isExpanded>
        Your payment solution has not been properly configured, please contact your Reapit Administrator if this problem
        persists.
      </PersistentNotification>
    )
  }

  return <PaymentPageContent paymentProvider={paymentProvider} />
}
