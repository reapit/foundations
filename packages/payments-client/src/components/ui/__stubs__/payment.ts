import { PaymentModel } from '@reapit/foundations-ts-definitions'

export const data: PaymentModel = {
  id: 'PAY20000001',
  created: '2018-12-12T12:30:23.0000000Z',
  modified: '2019-01-08T12:30:34.0000000Z',
  type: 'paymentRequest',
  description: 'Tenancy Check Fee',
  status: 'awaitingPosting',
  ledger: 'tenant',
  amount: 100,
  clientAccountName: 'Primary',
  externalReference: '',
  companyId: '',
  customer: {
    id: 'OXF12300101',
    name: 'Mr John Smith',
    title: 'Mr',
    forename: 'John',
    surname: 'Smith',
    type: 'contact',
    homePhone: '01234 567890',
    workPhone: '',
    mobilePhone: '07890 123456',
    email: 'example@email.com',
    primaryAddress: {
      buildingName: '',
      buildingNumber: '1',
      line1: 'Test House',
      line2: 'Test Lane',
      line3: 'County of Test',
      line4: 'Test City',
      postcode: 'A11 2BB',
      countryId: 'GB',
    },
  },
  landlordId: '',
  propertyId: 'PRP200001',
  tenancyId: '',
  _eTag: '',
  _links: {
    self: {
      href: '/payments/PAY20000001',
    },
    customer: {
      href: '/contacts/OXF12300101',
    },
    property: {
      href: '/properties/PRP200001',
    },
  },
}
