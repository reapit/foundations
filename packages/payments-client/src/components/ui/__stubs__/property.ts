import { PropertyModel } from '@reapit/foundations-ts-definitions'

export const data: PropertyModel = {
  id: 'OXF200008',
  created: '2018-12-12T12:30:23.0000000Z',
  modified: '2019-01-08T12:30:34.0000000Z',
  marketingMode: 'sellingAndLetting',
  currency: 'GBP',
  address: {
    buildingName: '',
    buildingNumber: '15',
    line1: 'Example street',
    line2: 'Solihull',
    line3: 'West Midlands',
    line4: '',
    postcode: 'B91 2XX',
    countryId: 'GB',
    geolocation: {
      latitude: 52.4158586,
      longitude: 1.7773383,
    },
  },
  areaId: 'BRM',
  strapline: '',
  description:
    'We are delighted to offer for sale this EXTENDED THREE BEDROOMED SEMI DETACHED PROPERTY situated in a much sought after residential location of Greasby, having the benefits of two separate entertaining rooms, morning room, extended kitchen. To the first floor there are three bedrooms, spacious family bathroom, gas central heating gardens front and rear and off road parking.',
  summary: '',
  departmentId: 'G',
  negotiatorId: 'JAS',
  bedrooms: 4,
  receptions: 1,
  bathrooms: 2,
  councilTax: 'A',
  internetAdvertising: true,
  viewingArrangements: 'Accompanied viewings after 3PM only',
  videoUrl: 'https://www.example.com/property/123/videoTour',
  videoCaption: 'Virtual property tour',
  externalArea: {
    type: 'acres',
    min: 1,
    max: 2,
  },
  internalArea: {
    type: 'squareFeet',
    min: 1500,
    max: 2000,
  },
  epc: {
    exempt: false,
    eer: 45,
    eerPotential: 71,
    eir: 53,
    eirPotential: 81,
  },
  selling: {
    instructed: '2018-11-18',
    price: 250000,
    qualifier: 'askingPrice',
    status: 'underOffer',
    disposal: 'privateTreaty',
    completed: '2019-08-27',
    exchanged: '2019-08-15',
    tenure: {
      type: 'leasehold',
      expiry: '2019-05-01',
    },
    vendorId: 'OXF123001',
  },
  letting: {
    instructed: '2018-11-18',
    availableFrom: '2018-12-26',
    availableTo: '2019-03-23',
    rent: 750,
    rentFrequency: 'monthly',
    furnishing: ['furnished', 'partFurnished'],
    term: 'long',
    status: 'toLet',
    agentRole: 'managed',
    landlordId: 'OXF123001',
  },
  type: ['house'],
  style: ['detached'],
  situation: ['garden', 'land', 'patio'],
  parking: ['secure', 'doubleGarage'],
  age: ['period'],
  locality: ['rural', 'village'],
  rooms: [
    {
      name: 'Kitchen',
      dimensions: '4.5m x 5.2m',
      description:
        'The breakfast kitchen, with utility, comprises a matching range of wall and base units with ceramic hob and eye level double oven. A side door leads to the front courtyard area.',
    },
    {
      name: 'Lounge',
      dimensions: '6.5m x 7.8m',
      description:
        'The lounge, with a brick feature fireplace housing a Clearview stove, provides an ideal space to relax and unwind, whilst enjoying views over the Golf Course.',
    },
    {
      name: 'Dining Room',
      dimensions: '5.1m x 6.2m',
      description: 'The formal dining room is the perfect space to entertain and provides access to the useful study.',
    },
  ],
  officeIds: ['OXF', 'SOL'],
  metadata: {
    CustomField1: 'CustomValue1',
    CustomField2: true,
  },
  _eTag: '',
  _links: {
    self: {
      href: '/properties/OXF200008',
    },
    images: {
      href: '/propertyImages/?propertyId=OXF200008',
    },
    documents: {
      href: '/documents/?associatedType=property&associatedId=OXF200008',
    },
    offers: {
      href: '/offers/?propertyId=OXF200008',
    },
    appointments: {
      href: '/appointments/?propertyId=OXF200008',
    },
    negotiator: {
      href: '/negotiators/JAS',
    },
    offices: {
      href: '/offices/?id=OXF&id=SOL',
    },
    department: {
      href: '/departments/G',
    },
    vendor: {
      href: '/vendors/OXF123001',
    },
    landlord: {
      href: '/landlords/OXF123001',
    },
    area: {
      href: '/areas/BRM',
    },
  },
}
