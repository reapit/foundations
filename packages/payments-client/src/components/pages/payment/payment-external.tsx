import React, { useState, useEffect, FC, SetStateAction, Dispatch } from 'react'
import { PaymentPageContent } from './payment-page-content'
import { PaymentWithPropertyModel } from '../../../types/payment'
import { handlePaymentProvider } from './payment-handlers'
import { Loader, PageContainer, PersistentNotification, useSnack } from '@reapit/elements'
import { PaymentParams, usePaymentsState } from '../../../core/use-payments-state'
import { getPaymentWithProperty } from '../../../services/payment'

export const handleSetPaymentWithProperty =
  (
    setPaymentWithProperty: Dispatch<SetStateAction<PaymentWithPropertyModel | null>>,
    setPaymentWithPropertyLoading: Dispatch<SetStateAction<boolean>>,
    paymentParams: PaymentParams,
    errorSnack: (message: string) => void,
  ) =>
  () => {
    const fetchPaymentWithProperty = async () => {
      const { paymentId, clientId, session } = paymentParams
      if (paymentId && clientId && session) {
        setPaymentWithPropertyLoading(true)

        const paymentWithPropertyModel = await getPaymentWithProperty(paymentParams, errorSnack)

        if (paymentWithPropertyModel && clientId) {
          setPaymentWithProperty({ ...paymentWithPropertyModel, clientCode: clientId })
        }

        setPaymentWithPropertyLoading(false)
      }
    }

    fetchPaymentWithProperty()
  }

export const PaymentExternalPage: FC = () => {
  const { paymentsDataState } = usePaymentsState()
  const [providerLoading, setProviderLoading] = useState(false)
  const [paymentWithPropertyLoading, setPaymentWithPropertyLoading] = useState(false)
  const { error: errorSnack } = useSnack()

  const { paymentProvider, paymentParams, paymentWithProperty, setPaymentProvider, setPaymentWithProperty } =
    paymentsDataState
  const { clientId } = paymentParams

  useEffect(handlePaymentProvider(setProviderLoading, setPaymentProvider, errorSnack, clientId), [
    setPaymentProvider,
    clientId,
  ])

  useEffect(
    handleSetPaymentWithProperty(setPaymentWithProperty, setPaymentWithPropertyLoading, paymentParams, errorSnack),
    [paymentParams],
  )

  if (providerLoading || paymentWithPropertyLoading) {
    return <Loader />
  }

  if (!paymentWithProperty) {
    return (
      <PageContainer>
        <PersistentNotification intent="secondary" isFullWidth isInline isExpanded>
          No valid session was found for this transaction. It is likely that the payment request has expired (refer to
          your email), or the payment has been made successfully. If you think you are seeing this message in error,
          please contact your agent so they can re-issue the payment request.
        </PersistentNotification>
      </PageContainer>
    )
  }

  if (!paymentProvider) {
    return (
      <PageContainer>
        <PersistentNotification intent="secondary" isFullWidth isInline isExpanded>
          There seems to be a problem with account with the payment provider. Please inform your agent of this issue and
          they will provide you with payment alternatives.
        </PersistentNotification>
      </PageContainer>
    )
  }

  return <PaymentPageContent />
}
