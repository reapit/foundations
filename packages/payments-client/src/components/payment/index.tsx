import React, { FC } from 'react'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { PaymentModel } from '@reapit/foundations-ts-definitions'
import {
  Button,
  elMb3,
  elMb5,
  FlexContainer,
  Icon,
  PageContainer,
  SecondaryNavContainer,
  SmallText,
  useModal,
} from '@reapit/elements'
import { GetActionNames, getActions, useReapitGet } from '@reapit/use-reapit-data'
import { useParams } from 'react-router-dom'
import { openNewPage } from '../../core/nav'
import PaymentRequestModal from '../payments/payment-request-modal'
import { useConfigState } from '../../core/use-config-state'
import { Payment } from './payment'

export const PaymentPage: FC = () => {
  const { paymentId } = useParams<'paymentId'>()
  const { Modal, openModal, closeModal } = useModal()
  const { config } = useConfigState()
  const configNotConfigured = !config?.isConfigured

  const paymentRequest = useReapitGet<PaymentModel>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getPaymentById],
    uriParams: {
      paymentId,
    },
    fetchWhenTrue: [paymentId],
  })

  const [payment, , , refreshPayment] = paymentRequest

  return (
    <FlexContainer isFlexAuto>
      <SecondaryNavContainer>
        <Icon className={elMb5} iconSize="large" icon="newCustomerInfographic" />
        <SmallText hasGreyText>
          From this page you can either send an email request for payment to the customer using the button below, or
          take a payment in person using our payment form.
        </SmallText>
        <Button className={elMb3} intent="neutral" onClick={openNewPage('')}>
          View Docs
        </Button>
        <Button intent="primary" disabled={payment?.status === 'posted' || configNotConfigured} onClick={openModal}>
          Email Request
        </Button>
      </SecondaryNavContainer>
      <PageContainer>
        <Payment paymentRequest={paymentRequest} />
      </PageContainer>
      <Modal title={`Request Payment of Â£${payment?.amount ? payment?.amount.toFixed(2) : 0} for ${payment?.id}`}>
        <PaymentRequestModal closeModal={closeModal} refreshPayments={refreshPayment} selectedPayment={payment} />
      </Modal>
    </FlexContainer>
  )
}

export default PaymentPage
