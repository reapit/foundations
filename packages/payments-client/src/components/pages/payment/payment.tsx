import React, { Dispatch, FC, SetStateAction, useEffect } from 'react'
import { Route, useLocation, useParams } from 'react-router'
import { Routes } from '../../../constants/routes'
import { PaymentExternalPage } from './payment-external'
import { PaymentInternalPage } from './payment-internal'
import { PaymentParams, usePaymentsState } from '../../../core/use-payments-state'

export interface PaymentUriParams {
  paymentId: string
}

export const handleSetUriParams =
  (
    setPaymentParams: Dispatch<SetStateAction<PaymentParams>>,
    paymentId: string,
    session: string | null,
    clientId: string | null,
  ) =>
  () => {
    setPaymentParams({ paymentId, session, clientId })
  }

export const Payment: FC = () => {
  const location = useLocation()
  const { paymentId } = useParams<PaymentUriParams>()
  const { paymentsDataState } = usePaymentsState()
  const { search } = location
  const queryParams = new URLSearchParams(search)
  const session = queryParams.get('session')
  const clientId = queryParams.get('clientCode')

  const { setPaymentParams } = paymentsDataState

  useEffect(handleSetUriParams(setPaymentParams, paymentId, session, clientId), [paymentId, session, clientId])

  if (session && clientId) {
    return <Route path={Routes.PAYMENT} component={PaymentExternalPage} exact />
  }

  return <Route path={Routes.PAYMENT} component={PaymentInternalPage} exact />
}

export default Payment
