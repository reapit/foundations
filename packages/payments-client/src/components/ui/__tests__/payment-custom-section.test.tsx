import * as React from 'react'
import { shallow } from 'enzyme'
import PaymentCustomerSection from '../payment-customer-section'
import { stubPaymentModel } from '../__stubs__/payment'

describe('PaymentCustomerSection', () => {
  it('should match a snapshot', () => {
    const customer = stubPaymentModel.customer
    const wrapper = shallow(<PaymentCustomerSection customer={customer} />)
    expect(wrapper).toMatchSnapshot()
  })
})
