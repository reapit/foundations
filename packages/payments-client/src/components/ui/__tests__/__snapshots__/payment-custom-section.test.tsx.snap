// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentCustomerSection should match a snapshot 1`] = `
<GridItem>
  <FadeIn>
    <H5>
      Customer
    </H5>
    <IconList
      items={
        Array [
          Object {
            "icon": <FaUserAlt
              className="icon-list-icon"
            />,
            "text": "Mr John Smith",
          },
          Object {
            "icon": <FaHome
              className="icon-list-icon"
            />,
            "text": "1 Test House Test Lane County of Test Test City A11 2BB",
          },
          Object {
            "icon": <FaEnvelope
              className="icon-list-icon"
            />,
            "text": "example@email.com",
          },
          Object {
            "icon": <FaStickyNote
              className="icon-list-icon"
            />,
            "text": "Customer ref: OXF12300101",
          },
        ]
      }
    />
  </FadeIn>
</GridItem>
`;
